### PR TITLE
IGNITE-18567 Configuration validators code refactored.

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-18567

### CODE REVIEW

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+     */
+    static boolean canValidate(
+            Validator<?, ?> validator,
+            Class<? extends Annotation> annotationType,
+            Class<?> schemaFieldType,
+            boolean namedList
+    ) {
+        Type[] genericTypeParameters = genericTypeParameters(validator);
+
+        if (genericTypeParameters[0] != annotationType) {
```
Comment (tkalkirill) [inline modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java:52]: Add some description with examples.

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+            return false;
+        }
+
+        Type viewType = genericTypeParameters(validator)[1];
```
Comment (tkalkirill) [inline modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java:56]: Why call `genericTypeParameters` again?

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+        if (viewType instanceof ParameterizedType) {
+            ParameterizedType parameterizedType = (ParameterizedType) viewType;
+
+            assert parameterizedType.getRawType() == NamedListView.class;
```
Comment (tkalkirill) [inline modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java:61]: Missed printout what's wrong.

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+                return false;
+            }
+
+            Type namedListElementType = parameterizedType.getActualTypeArguments()[0];
+
+            if (namedListElementType instanceof WildcardType) {
+                return true;
+            }
+
+            assert namedListElementType instanceof Class;
```
Comment (tkalkirill) [inline modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java:73]: Missed printout what's wrong.

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+            Class<?> namedListElementClass = (Class<?>) namedListElementType;
+
+            return canValidateConfigValue(schemaFieldType, namedListElementClass);
+        }
+
+        if (namedList) {
+            return false;
+        }
+
+        assert viewType instanceof Class;
```
Comment (tkalkirill) [inline modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java:84]: Missed printout what's wrong.

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/ConfigurationChanger.java [hunk 6]
```diff
@@ -157,13 +158,13 @@ private StorageRoots(SuperRoot roots, long version) {
...
             Collection<RootKey<?, ?>> rootKeys,
-            Map<Class<? extends Annotation>, Set<Validator<?, ?>>> validators,
+            Collection<Validator<?, ?>> validators,
             ConfigurationStorage storage
     ) {
         checkConfigurationType(rootKeys, storage);
 
         this.notificator = notificator;
-        this.validators = validators;
+        this.validators = new ArrayList<>(validators);
```
Comment (tkalkirill) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/ConfigurationChanger.java:167]: Maybe: `List.copyOf(validators)` ?

**Review by tkalkirill — APPROVED** [review]

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java (hunk 2 above)
Comment (ibessonov) [reply modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java]: Fixed

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java (hunk 1 above)
Comment (ibessonov) [reply modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java:54]: Fixed

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java (hunk 3 above)
Comment (ibessonov) [reply modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java]: Done

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java (hunk 4 above)
Comment (ibessonov) [reply modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java]: Done

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java (hunk 5 above)
Comment (ibessonov) [reply modules/configuration-api/src/main/java/org/apache/ignite/configuration/validation/ValidatorChecker.java]: Done

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/ConfigurationChanger.java (hunk 6 above)
Comment (ibessonov) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/ConfigurationChanger.java:167]: Sure

**Pull request merged**